daemon/daemon.go

```diff
@@ -521,6 +525,32 @@ func (d *Daemon) compileBase() error {
 		return err
 	}
 
+	// Validate alignments of C and Go equivalent structs
+	toCheck := map[string]reflect.Type{
+		"ipv4_ct_tuple":        reflect.TypeOf(ctmap.CtKey4{}),
+		"ipv6_ct_tuple":        reflect.TypeOf(ctmap.CtKey6{}),
+		"ct_entry":             reflect.TypeOf(ctmap.CtEntry{}),
+		"ipcache_key":          reflect.TypeOf(ipcachemap.Key{}),
+		"remote_endpoint_info": reflect.TypeOf(ipcachemap.RemoteEndpointInfo{}),
+		"lb4_key":              reflect.TypeOf(lbmap.Service4Key{}),
+		"lb4_service":          reflect.TypeOf(lbmap.Service4Value{}),
+		"lb6_key":              reflect.TypeOf(lbmap.Service6Key{}),
+		"lb6_service":          reflect.TypeOf(lbmap.Service6Value{}),
+		"endpoint_key":         reflect.TypeOf(bpf.EndpointKey{}),
+		"endpoint_info":        reflect.TypeOf(lxcmap.EndpointInfo{}),
+		"metrics_key":          reflect.TypeOf(metricsmap.Key{}),
+		"metrics_value":        reflect.TypeOf(metricsmap.Value{}),
+		"proxy4_tbl_key":       reflect.TypeOf(proxymap.Proxy4Key{}),
+		"proxy4_tbl_value":     reflect.TypeOf(proxymap.Proxy4Value{}),
+		"proxy6_tbl_key":       reflect.TypeOf(proxymap.Proxy6Key{}),
+		"proxy6_tbl_value":     reflect.TypeOf(proxymap.Proxy6Value{}),
+		"sock_key":             reflect.TypeOf(sockmap.SockmapKey{}),
+		"ep_config":            reflect.TypeOf(configmap.EndpointConfig{}),
+	}
+	if err := alignchecker.CheckStructAlignments("bpf_alignchecker.o", toCheck); err != nil {
```

Comment:
Why do we need to do this at runtime? Couldn't we do this in a unit test (which from the Makefile depends on a build target for the alignchecker binary under bpf/), or ginkgo test (similar to eg the verifier tests)?

Reply:
> Why do we need to do this at runtime?

Mostly because we compile BPF programs at runtime, and probably none of users runs unit/ginkgo tests.

I expect some users to run Cilium distributed not in a container image, so they might rely on their host LLVM suite which, at least in theory, can produce structs with different offsets than the ones we expect in Go.